waspc/data/Generator/templates/server/src/auth/providers/oauth/user.ts

```diff
@@ -0,0 +1,85 @@
+{{={= =}=}}
+
+import {
+  type ProviderId,
+  createUser,
+  sanitizeAndSerializeProviderData,
+  validateAndGetUserFields,
+  createProviderId,
+} from 'wasp/auth/utils'
+import { type {= authEntityUpper =} } from 'wasp/entities'
+import { prisma } from 'wasp/server'
+import { type UserSignupFields, type ProviderConfig } from 'wasp/auth/providers/types'
+import { getRedirectUriForOneTimeCode, getRedirectUriForError } from './redirect'
+import { tokenStore } from './oneTimeCode'
+
+export async function finishOAuthFlowAndGetRedirectUri(
+  provider: ProviderConfig,
+  providerProfile: unknown,
+  providerUserId: string,
+  userSignupFields?: UserSignupFields,
+): Promise<{ redirectUri: string }> {
+  const providerId = createProviderId(provider.id, providerUserId);
+
+  const authId = await getAuthIdFromProviderDetails(providerId, providerProfile, userSignupFields);
+
+  if (!authId) {
+      throw new Error("Something went wrong while creating the user. Please try again.");
+  }
+
+  const oneTimeCode = await tokenStore.createToken(authId);
+
+  return {
+      redirectUri: getRedirectUriForOneTimeCode(oneTimeCode),
+  };
+}
+
+export function handleOAuthErrorAndGetRedirectUri(error: unknown): { redirectUri: string } {
+  const errorMessage = (error as any).message ?? "An unknown error occurred while trying to log in with the OAuth provider.";
+  return {
+    redirectUri: getRedirectUriForError(errorMessage),
+  };
```

Comment:
How come you are wrapping this inside an object.

If it's because you want a smart constructor, it makes sense to extract the type (and employ a symbol, as mentioned in the other comments).

Still, I think it's OK if you just return a string.

Reply:
I've decided to return URL since Arctic's method for generating the login URL returns an actual `URL` object. It keeps things consistent and also validates the URL string which is a nice side-effect.

Reply:
Even better